### PR TITLE
feat(prometheus): add histogram

### LINKED ISSUE #1118: Add histogram metric on attestation reception

### PR DESCRIPTION
**Pull request by migmartri**
It adds Prometheus histogram with information about the workflow runs 

This histograms is recorded in the following places

- When an attestation is successfully stored
- Or canceled/error (reset)
- Or expired in the background

It includes the labels `org`, `runner type`, `status`, and `workflow` and it creates metric buckets that go from 1 second to 20 minutes. This can be tweaked later

```
curl -H "Authorization: Bearer $TOKEN"  localhost:8000/prom/development/metrics 
# HELP chainloop_wf_run_duration_seconds Duration of a workflow runs in seconds.
# TYPE chainloop_wf_run_duration_seconds histogram
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="10"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="30"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="60"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="90"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="120"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="180"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="240"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="300"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="600"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="900"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="1200"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="GITHUB_ACTION",status="success",workflow="foo",le="+Inf"} 1
chainloop_wf_run_duration_seconds_sum{org="development",runner="GITHUB_ACTION",status="success",workflow="foo"} 4.124993031
chainloop_wf_run_duration_seconds_count{org="development",runner="GITHUB_ACTION",status="success",workflow="foo"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="10"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="30"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="60"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="90"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="120"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="180"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="240"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="300"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="600"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="900"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="1200"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo",le="+Inf"} 1
chainloop_wf_run_duration_seconds_sum{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo"} 3.989257525
chainloop_wf_run_duration_seconds_count{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="canceled",workflow="foo"} 1
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="10"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="30"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="60"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="90"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="120"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="180"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="240"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="300"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="600"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="900"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="1200"} 2
chainloop_wf_run_duration_seconds_bucket{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo",le="+Inf"} 2
chainloop_wf_run_duration_seconds_sum{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo"} 6.257108346000001
chainloop_wf_run_duration_seconds_count{org="development",runner="RUNNER_TYPE_UNSPECIFIED",status="success",workflow="foo"} 2

```

Closes #1118

### CODE REVIEW

File: app/controlplane/internal/usercontext/apitoken_middleware.go [hunk 1]
```diff
@@ -138,6 +138,11 @@ func WithAttestationContextFromAPIToken(apiTokenUC *biz.APITokenUseCase, logger
 				return nil, fmt.Errorf("error extracting organization from APIToken: %w", err)
 			}
 
+			ctx, err = setCurrentOrgAndAPIToken(ctx, apiTokenUC, orgUC, tokenID)
```
Comment (migmartri) [inline app/controlplane/internal/usercontext/apitoken_middleware.go:141]: setting current org was missing in this middleware

File: app/controlplane/internal/server/grpc.go [hunk 2]
```diff
@@ -212,7 +212,7 @@ func craftMiddleware(opts *Opts) []middleware.Middleware {
 			// 2.a - Set its workflow and organization in the context
 			usercontext.WithAttestationContextFromRobotAccount(opts.RobotAccountUseCase, opts.OrganizationUseCase, logHelper),
 			// 2.b - Set its API token and Robot Account as alternative to the user
-			usercontext.WithAttestationContextFromAPIToken(opts.APITokenUseCase, logHelper),
+			usercontext.WithAttestationContextFromAPIToken(opts.APITokenUseCase, opts.OrganizationUseCase, logHelper),
```
Comment (migmartri) [inline app/controlplane/internal/server/grpc.go:215]: this is something that we were missing in our middleware. Technically it's not from this patch. I needed it in previous implementations of it. If you want I could try to move it outside.

File: app/controlplane/pkg/biz/workflowrun.go [hunk 3]
```diff
@@ -110,9 +115,9 @@ type WorkflowRunExpirerOpts struct {
 	CheckInterval    time.Duration
 }
 
-func NewWorkflowRunExpirerUseCase(wfrRepo WorkflowRunRepo, logger log.Logger) *WorkflowRunExpirerUseCase {
+func NewWorkflowRunExpirerUseCase(wfrRepo WorkflowRunRepo, po PromObservable, logger log.Logger) *WorkflowRunExpirerUseCase {
```
Comment (migmartri) [inline app/controlplane/pkg/biz/workflowrun.go:118]: I am passing an interface and not an actual type here so it helps with testing.

File: app/controlplane/pkg/data/workflowrun.go [hunk 4]
```diff
@@ -190,6 +190,7 @@ func (r *WorkflowRunRepo) ListNotFinishedOlderThan(ctx context.Context, olderTha
 	// TODO: Look into adding upper bound on the createdAt column to prevent full table scans
 	// For now this is fine especially because we have a composite index
 	workflowRuns, err := r.data.DB.WorkflowRun.Query().
+		WithWorkflow().
```
Comment (migmartri) [inline app/controlplane/pkg/data/workflowrun.go:193]: we need the workflow to extract the `orgID`

File: app/controlplane/pkg/metrics/prometheus/registry/registry.go [hunk 5]
```diff
@@ -28,19 +28,34 @@ type PrometheusRegistry struct {
 	*prometheus.Registry
 	Name               string
 	chainloopCollector *collector.ChainloopCollector
+
+	// metrics
+	WorkflowRunDurationSeconds *prometheus.HistogramVec
 }
 
+var workflowRunDurationSeconds = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Name: "chainloop_wf_run_duration_seconds",
```
Comment (javirln) [inline app/controlplane/pkg/metrics/prometheus/registry/registry.go:37]: In the case of the already existing metric, it's called: `chainloop_workflow_up`, in order to be consistent with the naming, shall we name this `chainloop_workflow_run_duration_seconds` ?

I don't think we should be mixing short names with extended ones.

Comment (migmartri) [reply app/controlplane/pkg/metrics/prometheus/registry/registry.go]: Done, ptal

File: app/controlplane/internal/server/grpc.go (hunk 2 above)
Comment (javirln) [reply app/controlplane/internal/server/grpc.go:215]: I think it's fine here, no worries, thanks!!

**Pull request merged**